File: clients/capacitor/android/app/src/main/java/org/ubersdr/mobile/UberSdrPlugin.java
package org.ubersdr.mobile;

import android.content.Intent;
import android.net.Uri;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.io.IOException;
import java.util.List;

/**
 * The chooser's platform half: everything window.ubersdr needs that a WebView
 * cannot do.
 *
 * <p>The desktop client answers the same surface from 32 IPC handlers in its
 * main process (clients/electron/main.js). Rather less is needed here, because
 * the store, the discovery logic and the row shapes stayed in JavaScript — see
 * src/api.js. What is left is what genuinely needs the platform: HTTP with
 * per-receiver certificate trust, the local network, the keystore, and opening
 * a receiver.
 */
@CapacitorPlugin(name = "UberSdr")
public class UberSdrPlugin extends Plugin {

    private static UberSdrPlugin instance;

    private Secrets secrets;
    private LocalProxy proxy;
    private String openId;
    // Which receiver launch is the current one. See receiverClosed.
    private int epoch;

    @Override
    public void load() {
        instance = this;
        secrets = new Secrets(getContext());
    }

    // --- deep links ----------------------------------------------------------

    /** Whether a ubersdr:// link started or reached this run. */
    private boolean linkSeen = false;

    /**
     * A followed ubersdr:// link, on its way to src/deeplink.js.
     *
     * <p>Capacitor calls this for the intent that started the Activity as well
     * as for one delivered to it while it was running (BridgeActivity.load ends
     * with onNewIntent(getIntent())), so cold start and warm start are the same
     * path here. They differ in who is listening: on a cold start the page has
     * not been parsed yet, let alone registered anything, which is what
     * retainUntilConsumed is for — the event waits in the plugin until the
     * listener exists and is delivered then.
     *
     * <p>The one intent deliberately ignored is a relaunch from the recents
     * list. Android re-delivers the original VIEW intent when the app is
     * resumed that way after its process has gone, and following it would mean
     * a receiver the operator disconnected from yesterday reconnecting itself
     * today because the link that first opened it is still attached to the task.
     */
    @Override
    protected void handleOnNewIntent(Intent intent) {
        super.handleOnNewIntent(intent);
        if (intent == null) return;
        if ((intent.getFlags() & Intent.FLAG_ACTIVITY_LAUNCHED_FROM_HISTORY) != 0) return;

        Uri uri = intent.getData();
        if (uri == null || !"ubersdr".equalsIgnoreCase(uri.getScheme())) return;

        // Consumed: the same Intent object is what getIntent() keeps returning,
        // so clearing it stops a second delivery within this Activity's life
        // from being read as a second tap.
        intent.setData(null);

        // Remembered for the chooser, which has an "open the last receiver on
        // launch" setting and must stand aside for a link: the link names a
        // receiver somebody asked for now, where the setting names one they
        // asked for last time. Never cleared — the setting it guards is only
        // ever read once, as the page starts.
        linkSeen = true;

        JSObject data = new JSObject();
        data.put("url", uri.toString());
        notifyListeners("deepLink", data, true);
    }

    /**
     * Has a link been followed this run?
     *
     * <p>Answered from the Intent rather than from whether the page has
     * received the event yet, which is the whole point: on a cold start the two
     * race, and a chooser that asked "is a receiver open?" would find the
     * answer no while a directory lookup was still in flight and open a
     * different one.
     */
    @PluginMethod
    public void linkPending(PluginCall call) {
        JSObject out = new JSObject();
        out.put("pending", linkSeen);
        call.resolve(out);
    }

    // --- HTTP ----------------------------------------------------------------

    @PluginMethod
    public void getJson(PluginCall call) {
        final String host = call.getString("host", "");
        final int port = call.getInt("port", 0);
        final boolean tls = Boolean.TRUE.equals(call.getBoolean("tls", false));
        final boolean insecure = Boolean.TRUE.equals(call.getBoolean("insecureTLS", false));
        final String path = call.getString("path", "/");
        final int timeoutMs = call.getInt("timeoutMs", 8000);
        // From src/useragent.js, which owns the token — one spelling of this
        // client's name, shared by the probes and the receiver's WebView.
        final String userAgent = call.getString("userAgent", "UberSDR-Android");

        if (host.isEmpty() || port <= 0) {
            call.reject("a host and port are required");
            return;
        }

        // Off the main thread: this is a network call, and Capacitor runs
        // plugin methods on the caller's thread.
        getBridge().execute(() -> {
            Http.Result result = Http.getJson(host, port, tls, insecure, path, timeoutMs, userAgent);
            JSObject out = new JSObject();
            out.put("ok", result.ok);
            out.put("status", result.status);
            if (result.ok) {
                out.put("body", result.body);
            } else {
                out.put("code", result.code);
                out.put("error", result.error);
                out.put("certError", result.certError);
            }
            // Resolved either way. A receiver that is not there is an answer,
            // not an exception, and the discovery layer reads the fields.
            call.resolve(out);
        });
    }

    // --- the local network ---------------------------------------------------

    @PluginMethod
    public void mdnsBrowse(PluginCall call) {
        final int timeoutMs = call.getInt("timeoutMs", 3000);
        getBridge().execute(() -> {
            JSArray services = new JSArray();
            for (Mdns.Service svc : Mdns.browse(timeoutMs)) {
                JSObject o = new JSObject();
                o.put("name", svc.name);
                o.put("host", svc.host);
                o.put("port", svc.port);
                services.put(o);
            }
            JSObject out = new JSObject();
            out.put("services", services);
            call.resolve(out);
        });
    }

    // --- the bypass password -------------------------------------------------
    //
    // Set, cleared and asked about. There is no method that returns one: the
    // receiver Activity reads it from Secrets directly when it opens a page, so
    // a value the chooser must not hold never passes through JavaScript.

    @PluginMethod
    public void secretSet(PluginCall call) {
        String id = call.getString("id", "");
        if (id.isEmpty()) { call.reject("an id is required"); return; }
        secrets.set(id, call.getString("value", ""));
        call.resolve();
    }

    @PluginMethod
    public void secretHas(PluginCall call) {
        JSObject out = new JSObject();
        out.put("has", secrets.has(call.getString("id", "")));
        call.resolve(out);
    }

    @PluginMethod
    public void secretClear(PluginCall call) {
        secrets.clear(call.getString("id", ""));
        call.resolve();
    }

    // --- receivers -----------------------------------------------------------

    /**
     * Start this instance's loopback proxy and open a receiver on it.
     *
     * <p>Returns the port actually bound, which is the stored one unless
     * something else had it. The caller persists what comes back: the port is
     * the origin, and the origin is where the v2 UI keeps this receiver's
     * settings.
     */
    @PluginMethod
    public void openReceiver(PluginCall call) {
        final String id = call.getString("id", "");
        final String host = call.getString("host", "");
        final int port = call.getInt("port", 0);
        final boolean tls = Boolean.TRUE.equals(call.getBoolean("tls", false));
        final boolean insecure = Boolean.TRUE.equals(call.getBoolean("insecureTLS", false));
        final int preferredPort = call.getInt("localPort", 0);
        final String label = call.getString("label", "UberSDR");
        final String product = call.getString("product", "");

        if (id.isEmpty() || host.isEmpty() || port <= 0) {
            call.reject("an id, host and port are required");
            return;
        }

        // This receiver is already open: bring it forward and leave it alone.
        //
        // Restarting the proxy under a page that is using it would drop every
        // socket the receiver has — the audio, the spectrum, the session — and
        // then the page would reconnect through the new one, so the visible
        // effect of asking for what is already there would be a gap in the
        // audio. Two callers reach this: the chooser's button, which says
        // "Show" rather than "Connect" for a running receiver, and a link
        // followed for the receiver that is already playing.
        //
        // The Activity is singleTask, so an Intent carrying no id comes out at
        // its onNewIntent, which ignores it and simply arrives in front.
        if (proxy != null && id.equals(openId) && ReceiverActivity.isOpen()) {
            getActivity().startActivity(new Intent(getContext(), ReceiverActivity.class));
            JSObject shown = new JSObject();
            shown.put("localPort", proxy.localPort());
            call.resolve(shown);
            return;
        }

        // One receiver at a time. A phone shows one thing, and two receivers
        // playing at once would be two claims on the audio focus.
        stopProxy();

        LocalProxy started = new LocalProxy(host, port, tls, insecure, getContext().getAssets());
        int bound;
        try {
            bound = started.start(preferredPort);
        } catch (IOException e) {
            call.reject("could not start the local proxy: " + e.getMessage());
            return;
        }
        proxy = started;
        openId = id;
        epoch++;

        Intent intent = new Intent(getContext(), ReceiverActivity.class);
        intent.putExtra(ReceiverActivity.EXTRA_ID, id);
        intent.putExtra(ReceiverActivity.EXTRA_EPOCH, epoch);
        intent.putExtra(ReceiverActivity.EXTRA_URL, started.origin() + "/v2/");
        intent.putExtra(ReceiverActivity.EXTRA_ORIGIN, started.origin());
        intent.putExtra(ReceiverActivity.EXTRA_UPSTREAM, started.upstreamOrigin());
        intent.putExtra(ReceiverActivity.EXTRA_LABEL, label);
        intent.putExtra(ReceiverActivity.EXTRA_INSECURE, insecure);
        intent.putExtra(ReceiverActivity.EXTRA_PRODUCT, product);
        // Whose settings this receiver reads and writes. Decided in the
        // chooser and carried in, because the page is seeded before its first
        // script runs and there is nothing to ask by then.
        intent.putExtra(ReceiverActivity.EXTRA_PREFS_SCOPE, call.getString("prefsScope"));
        getActivity().startActivity(intent);

        JSObject out = new JSObject();
        out.put("localPort", bound);
        call.resolve(out);
    }

    @PluginMethod
    public void closeReceiver(PluginCall call) {
        // The Activity's onDestroy calls receiverClosed, which stops the proxy
        // — the same path the back gesture takes, so there is one way a
        // receiver ends rather than two that must agree.
        ReceiverActivity.finishCurrent();
        call.resolve();
    }

    /**
     * Throw away the interface settings the receivers share.
     *
     * <p>Only what this app stored on the page's behalf — the v2 snapshot in
     * Prefs. Not the saved receivers, not their passwords, and not what a
     * receiver keeps in its own localStorage: those belong to the chooser and
     * to the page, and a "reset settings" that quietly took the receiver list
     * with it would be a button nobody could risk pressing.
     */
    /** One app-level setting, for the chooser's settings page. See Prefs. */
    @PluginMethod
    public void prefGet(PluginCall call) {
        JSObject out = new JSObject();
        out.put("value", Prefs.readOne(getContext(), call.getString("key")));
        call.resolve(out);
    }

    @PluginMethod
    public void prefSet(PluginCall call) {
        Prefs.writeOne(getContext(), call.getString("key"), call.getString("value"));
        call.resolve();
    }

    @PluginMethod
    public void resetPrefs(PluginCall call) {
        Prefs.resetAll(getContext());
        call.resolve();
    }

    /**
     * This app's page in Android's settings.
     *
     * <p>Notifications are the reason it exists: a refusal is remembered, and
     * from Android 13 a second request is not even shown — so an operator who
     * said no has no way back from inside the app, and no way of knowing that
     * is where the answer lives.
     */
    @PluginMethod
    public void openAppSettings(PluginCall call) {
        try {
            Intent intent = new Intent(android.provider.Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
            intent.setData(android.net.Uri.fromParts("package", getContext().getPackageName(), null));
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            getContext().startActivity(intent);
            call.resolve();
        } catch (Exception e) {
            call.reject("no settings page on this device", e);
        }
    }

    @PluginMethod
    public void receiverState(PluginCall call) {
        JSObject out = new JSObject();
        out.put("id", openId);
        call.resolve(out);
    }

    /**
     * Called by ReceiverActivity.onDestroy, however it came to be destroyed.
     *
     * <p>The epoch is which launch that Activity belonged to, and is the whole
     * reason it exists: an Activity on its way out can report after the next one
     * has started. Following a link for the receiver that is already open does
     * exactly that — MainActivity is singleTask, so the launch itself destroys
     * the Activity above it, and the connect that follows can start a new proxy
     * before that Activity's onDestroy has run. Comparing ids would then have
     * this stop the proxy it had just been given, for a receiver that is
     * starting rather than one that has ended.
     */
    static void receiverClosed(String id, int closingEpoch) {
        UberSdrPlugin self = instance;
        if (self == null) return;
        if (closingEpoch == self.epoch) {
            self.stopProxy();
            self.openId = null;
        }
        JSObject data = new JSObject();
        data.put("id", id);
        self.notifyListeners("receiverClosed", data);
    }

    private void stopProxy() {
        if (proxy != null) {
            proxy.stop();
            proxy = null;
        }
    }

    @Override
    protected void handleOnDestroy() {
        stopProxy();
        instance = null;
        super.handleOnDestroy();
    }
}

File: clients/capacitor/ios/UberSdrPlugin/sources/UberSdrPlugin.swift
import Foundation
import Capacitor
import UserNotifications

/// The bridge the chooser talks to: `window.ubersdr` on this platform.
///
/// The iOS half of `UberSdrPlugin.java`, method for method, because the
/// JavaScript that calls it is the same JavaScript — `src/api.js` and
/// `src/discovery.js` are staged into both clients and neither knows which
/// platform answered. The contract is `src/native.js`:
///
///   getJson       one HTTP path for every call the discovery layer makes
///   mdnsBrowse    _ubersdr._tcp on the local network
///   secretSet / secretHas / secretClear
///                 the bypass password. There is no `secretGet` here, exactly
///                 as there is none on Android: the receiver reads it natively
///                 and the chooser never holds one.
///   openReceiver / closeReceiver / receiverState
///                 which receiver is open, if any
///
/// Registered by conforming to `CAPBridgedPlugin`, which Capacitor 7 discovers
/// through the Objective-C runtime — so a plugin living in a pod is found the
/// same way one in the app target would be, and this client's Swift can be
/// written without opening Xcode. See the podspec.
@objc(UberSdrPlugin)
public class UberSdrPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "UberSdrPlugin"
    public let jsName = "UberSdr"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getJson", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "mdnsBrowse", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "secretSet", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "secretHas", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "secretClear", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openReceiver", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "closeReceiver", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "receiverState", returnType: CAPPluginReturnPromise)
    ]

    /// Which receiver is open, if any. Nil until `openReceiver`, and back to nil
    /// when the receiver's own power button finishes its screen.
    private var openInstanceId: String?

    /// The screen itself, held so `closeReceiver` has something to dismiss.
    private var receiver: ReceiverViewController?

    // MARK: - ubersdr://connect?uuid=…

    /// A receiver, named in a link, from anywhere the phone can follow one — a
    /// QR code beside a radio, a message, an instance's own page.
    ///
    /// Capacitor turns `application(_:open:options:)` into a notification, so
    /// this needs no AppDelegate edit — which matters, because AppDelegate.swift
    /// is in the App target and this client's Swift deliberately all lives in a
    /// pod. See the podspec.
    ///
    /// Raised with `retainUntilConsumed` for the same reason as on Android: a
    /// link that *started* the app arrives before src/deeplink.js is listening,
    /// and an event delivered to nobody is a receiver that never opens.
    override public func load() {
        NotificationCenter.default.addObserver(
            forName: .capacitorOpenURL, object: nil, queue: .main
        ) { [weak self] notification in
            guard let object = notification.object as? [String: Any],
                  let url = object["url"] as? URL else { return }
            self?.handle(url: url)
        }

        #if DEBUG
        // A link, without a link to tap.
        //
        // `simctl openurl` reaches the app only through SpringBoard's "Open
        // in UberSDR?" alert, and simctl has no way to answer it — so on a
        // simulator there is otherwise no path to a receiver that does not go
        // through a human finger. This takes the same URL from the launch
        // environment instead:
        //
        //   SIMCTL_CHILD_UBERSDR_OPEN='ubersdr://connect?uuid=…' \
        //     xcrun simctl launch booted org.ubersdr.mobile
        //
        // It is the *same* path a real link takes — deeplink.js cannot tell the
        // difference — so what it exercises is the real one. DEBUG only: a
        // release build has no way to be told where to connect at launch.
        if let raw = ProcessInfo.processInfo.environment["UBERSDR_OPEN"],
           let url = URL(string: raw) {
            DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) { [weak self] in
                self?.handle(url: url)
            }
        }
        #endif
    }

    private func handle(url: URL) {
        // The scheme alone is registered, not scheme + host, so what a link
        // asks for is src/deeplink.js's to read and to refuse: a mistyped link
        // gets a sentence rather than the OS quietly offering the browser
        // instead, and a second kind of link later is a change to one file.
        guard url.scheme?.lowercased() == "ubersdr" else { return }
        notifyListeners("deepLink", data: ["url": url.absoluteString], retainUntilConsumed: true)
    }

    // MARK: - HTTP

    /// Resolves with a result rather than rejecting, even for a failure.
    ///
    /// That is not laziness about error handling, it is the contract
    /// `discovery.js` is written against: a receiver that is simply not there
    /// has to travel the bridge the same way as one that is, carrying `code`
    /// and `certError`, because `resolveTarget` reads those to decide whether to
    /// offer "trust this receiver anyway". A rejection would arrive as an
    /// exception with a string and none of that.
    @objc func getJson(_ call: CAPPluginCall) {
        guard let host = call.getString("host"), let path = call.getString("path") else {
            call.reject("host and path are required")
            return
        }
        Http.getJson(
            host: host,
            port: call.getInt("port") ?? 443,
            tls: call.getBool("tls") ?? false,
            insecureTLS: call.getBool("insecureTLS") ?? false,
            path: path,
            timeoutMs: call.getInt("timeoutMs") ?? 8000,
            userAgent: call.getString("userAgent") ?? "UberSDR-iOS"
        ) { result in
            call.resolve(result.payload)
        }
    }

    // MARK: - The local network

    @objc func mdnsBrowse(_ call: CAPPluginCall) {
        Mdns.browse(timeoutMs: call.getInt("timeoutMs") ?? 2500) { found in
            // `services`, and the name is not free to choose: discovery.js does
            // `const { services = [] } = await UberSdr.mdnsBrowse(...)`, so any
            // other key destructures to the default and the LAN tab shows
            // nothing at all — with no error, because an empty network is a
            // perfectly ordinary answer. The Android plugin returns the same
            // key for the same reason.
            call.resolve(["services": found.map { $0.payload }])
        }
    }

    // MARK: - The password

    @objc func secretSet(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), let value = call.getString("value") else {
            call.reject("id and value are required")
            return
        }
        call.resolve(["ok": Secrets.set(id, value)])
    }

    @objc func secretHas(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("id is required")
            return
        }
        call.resolve(["has": Secrets.has(id)])
    }

    @objc func secretClear(_ call: CAPPluginCall) {
        guard let id = call.getString("id") else {
            call.reject("id is required")
            return
        }
        call.resolve(["ok": Secrets.clear(id)])
    }

    // MARK: - The receiver

    /// Start this receiver's proxy and put its screen on top.
    ///
    /// Resolves with the port that was actually bound, which is not always the
    /// one that was asked for — see LocalProxy.start. api.js writes back
    /// whatever comes out, because that port is the receiver's origin and so
    /// where its settings live.
    @objc func openReceiver(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), let host = call.getString("host") else {
            call.reject("id and host are required")
            return
        }
        let proxy = LocalProxy(host: host,
                               port: call.getInt("port") ?? 443,
                               tls: call.getBool("tls") ?? false,
                               insecureTLS: call.getBool("insecureTLS") ?? false)
        let bound: Int
        do {
            bound = try proxy.start(preferredPort: call.getInt("localPort") ?? 0)
        } catch {
            call.reject("could not start the local proxy: \(error.localizedDescription)")
            return
        }

        let label = call.getString("label") ?? ""
        let product = call.getString("product") ?? "UberSDR-iOS"
        // Read here rather than handed over the bridge: the JavaScript half of
        // this app has no call that returns a password, exactly as on Android.
        let password = Secrets.get(id)

        // What the OS has already decided about notifications, before the page
        // is built: the seed script has to carry it, and asking is async. The
        // Android client reads the same thing synchronously with
        // checkSelfPermission — see ReceiverActivity.notificationState.
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            let state: String
            switch settings.authorizationStatus {
            case .authorized, .provisional, .ephemeral: state = "granted"
            case .denied:                                state = "denied"
            default:                                     state = "default"
            }

        DispatchQueue.main.async { [weak self] in
            guard let self = self,
                  let root = self.bridge?.viewController else {
                proxy.stop()
                call.reject("no view controller to present from")
                return
            }
            // A receiver open already. On Android the receiver is a singleTask
            // Activity and the platform sorts this out; here it would be a
            // modal presented over a modal, which UIKit refuses outright — so
            // the one that is up is stopped and dismissed first, and only then
            // is the new one presented. Reachable from a followed link while
            // listening, which is exactly what `ubersdr://` links are for.
            if let current = self.receiver {
                current.close()
                current.dismiss(animated: false)
                self.receiver = nil
            }

            let receiver = ReceiverViewController(instanceId: id, label: label,
                                                  proxy: proxy, product: product,
                                                  password: password,
                                                  notificationState: state)
            receiver.modalPresentationStyle = .fullScreen
            receiver.onClosed = { [weak self] in
                guard let self = self else { return }
                self.openInstanceId = nil
                PlaybackSession.end()
                // The chooser redraws its rows on this. See api.js.
                self.notifyListeners("receiverClosed", data: [:])
            }
            self.receiver = receiver
            self.openInstanceId = id

            // Claim the audio session before the page starts making noise, so
            // that playback continues when the screen locks. UIBackgroundModes
            // in Info.plist is the other half of this.
            PlaybackSession.begin()

            root.present(receiver, animated: true) {
                call.resolve(["localPort": bound])
            }
        }
        }
    }

    @objc func closeReceiver(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            if let receiver = self.receiver {
                receiver.close()
                receiver.dismiss(animated: true)
                self.receiver = nil
            }
            self.openInstanceId = nil
            PlaybackSession.end()
            call.resolve()
        }
    }

    @objc func receiverState(_ call: CAPPluginCall) {
        // `id: null` is the shape api.js expects for "nothing is open" — it
        // destructures `{ id }` and compares it, so the key must be present.
        call.resolve(["id": openInstanceId as Any])
    }
}
